sockeye/encoder.py

```diff
@@ -143,20 +147,38 @@ def __init__(self,
                 self.embed_weight = embed_weight  # adds to self._reg_params
                 self.params.update({embed_weight.name: embed_weight})  # adds to self.params
 
-            self.factor_embeds = None
             if self.config.factor_configs is not None:
-                self.factor_embeds = mx.gluon.nn.HybridSequential()
-                # Factor weights aren't shared so they're not passed in and we create them here.
-                for i, fc in enumerate(self.config.factor_configs, 1):
-                    self.factor_embeds.add(mx.gluon.nn.Embedding(fc.vocab_size, fc.num_embed,
-                                                                 prefix="factor%d_" % i))
-
-    def hybrid_forward(self, F, data, valid_length, embed_weight):  # pylint: disable=arguments-differ
-        factor_embeds = []
+                for i, fc in enumerate(self.config.factor_configs):
+                    factor_weight_name = 'factor%d_weight' % i
+                    factor_weight = embed_weight if fc.share_source_embedding else \
+                        self.params.get('factor%d_weight' % i, shape=(fc.vocab_size, fc.num_embed))
+                    # We set the attribute of the class to trigger the hybrid_forward parameter creation "magic"
```

Comment:
What is the magic here?

Reply:
mxnet defines the parameters of the block (and thus passed to hybrid_forward) as the members of the block which are of class Parameter. The goal is to use something like
```
self.my_param = self.params.get('blablabla', ...)
```
But in our case we create the parameters dynamically. Simply storing them in a list does not work (`self.factor_weights = [self.params.get(...) for ...]`), as `factor_weights` would be of type List. Therefore we create a number of class attributes of the appropriate class.